### DISCUSSION TITLE
Microsoft.Extensions.Http.Resilience support for .NET Framework

### DISCUSSION THREAD

**Discussion by ayaya-dev** [dotnet/extensions#4952] [Q&A] [answered]
Correct me if this is the wrong place. But since Microsoft.Extensions.Http.Resilience is a replacement for Microsoft.Extensions.Http.Polly we tried updating to it in our .NET Framework 4.8 project. However, this gives us a plethora of warnings for various packages.

> Microsoft.Extensions.AmbientMetadata.Application 8.2.0 doesn't support net48 and has not been tested with it. Consider upgrading your TargetFramework to net6.0 or later. You may also set <SuppressTfmSupportBuildWarnings>true</SuppressTfmSupportBuildWarnings> in the project file to ignore this warning and attempt to run in this unsupported configuration at your own risk.

Maybe I'm missing something but the projects that get's the warning should support .NET 4.8 according to nuget.org.

Is Microsoft.Extensions.Http.Resilience supposed to support .NET Framework or is it just a mistake?

**Comment by RussKie** [comment, marked as answer]
No, it is not a mistake. 

While the carried binaries compatible with the .NET Framework 4.6.2 onwards, Microsoft.Extensions.Http.Resilience (or any other Microsoft.Extensions.* packages for that matter) is not officially supported. The .NET Framework binaries are there to support Microsoft internal teams for the period of migration to .NET platform.

FWIW, Microsoft.Extensions.* packages originating from dotnet/runtime repo have the same warnings.

I understand this is probably not the answer you expected but this is the reality. That said, you can suppress the build warning `<SuppressTfmSupportBuildWarnings>true</SuppressTfmSupportBuildWarnings>` in your proj (or the global props) file and consume the package as before.

**Reply by KalleOlaviNiemitalo** [reply]
> FWIW, Microsoft.Extensions.* packages originating from dotnet/runtime repo have the same warnings.

Microsoft.Extensions.Configuration.Binder warns about .NET Framework 4.6.1 compatibility and suggests targeting .NET Framework 4.6.2 instead.  This doesn't seem to be the same policy as in the dotnet/extensions repository.

<https://github.com/dotnet/runtime/blob/2f43856e78faebe7347fdd76acc26de7ffaec539/src/libraries/Microsoft.Extensions.Configuration.Binder/src/buildTransitive/Microsoft.Extensions.Configuration.Binder.targets#L16-L30>

**Reply by KalleOlaviNiemitalo** [reply]
See also <https://github.com/dotnet/extensions-samples/issues/217#issuecomment-1888066241> about the .NET Framework support difference between dotnet/runtime and dotnet/extensions.

**Reply by ayaya-dev** [reply]
Thanks for the clarification! We do not wish to enable SuppressTfmSupportBuildWarnings for all projects so I think we'll have to skip using the package. I do want to clarify though that the packages from dotnet/runtime does not have the same warning when using net48.

Maybe it should be documented somewhere for users that this is the way it is? At least I it made me confused since the warning says it doesn't support net48 while the package details says it does.

**Reply by KalleOlaviNiemitalo** [reply]
If you want to use the dotnet/extensions packages in the "unsupported configuration" and suppress the warnings for these packages only, you can add a Directory.Build.targets file that overrides the warning targets defined in the packages:

```XML
<Project>
  <Target Name="NETStandardCompatError_Microsoft_Extensions_AmbientMetadata_Application_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Compliance_Abstractions_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_DependencyInjection_AutoActivation_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Diagnostics_ExceptionSummarization_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Http_Diagnostics_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Http_Resilience_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Resilience_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Telemetry_net6_0" />
  <Target Name="NETStandardCompatError_Microsoft_Extensions_Telemetry_Abstractions_net6_0" />
</Project>
```

No guarantees that this will keep working in future versions.

**Reply by KalleOlaviNiemitalo** [reply]
> Maybe it should be documented somewhere for users that this is the way it is? 

The README file of each package would be the obvious place to add this.

**Reply by joperezr** [reply]
> FWIW, Microsoft.Extensions.* packages originating from dotnet/runtime repo have the same warnings.

Small clarification here, but packages from dotnet/runtime does support .NET Framework and don't have the same warnings. Their warnings, are when you try to run on out-of-support platforms, like .NET Core 2.0, 2.1, etc.

**Reply by RichardD2** [reply]
So is that "not officially supported" in the sense that *"it should work, but if it breaks, you're on your own"*? Or is it actively not going to work on .NET Framework?

We can't migrate to .NET, since we rely on SSRS support, which Microsoft has consistently refused to even consider adding. The current status of that seems to be *"move all your data to Azure, pay for PowerBI premium, and rewrite all of your reports in PowerBI"*, which is not an option.

**Reply by RussKie** [reply]
In the sense that "_it should work, but if it breaks, you're on your own_".

**Reply by zhaparoff** [reply]
> If you want to use the dotnet/extensions packages in the "unsupported configuration" and suppress the warnings for these packages only, you can add a Directory.Build.targets file that overrides the warning targets defined in the packages:
> 
> ```
> <Project>
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_AmbientMetadata_Application_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Compliance_Abstractions_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_DependencyInjection_AutoActivation_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Diagnostics_ExceptionSummarization_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Http_Diagnostics_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Http_Resilience_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Resilience_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Telemetry_net6_0" />
>   <Target Name="NETStandardCompatError_Microsoft_Extensions_Telemetry_Abstractions_net6_0" />
> </Project>
> ```
> 
> No guarantees that this will keep working in future versions.

Just in case someone is looking for the same workaround with .NET 8, target names are ending with `_net8_0`

**Comment by zhaparoff** [comment]
Considering that project(s), this issue referring to, contain `netstandard2.0` as one of their targets and can be built for that target
https://github.com/dotnet/extensions/blob/edf9088ebb507658c60e964ebd5548421c5dbaea/src/Libraries/Microsoft.Extensions.Http.Resilience/Microsoft.Extensions.Http.Resilience.csproj#L2C1-L7C19
they **should** be compatible with `net48`, at least at the source code level, the same way how the rest of extension packages are (Logging, DI, etc.).
So alternative solution could be making a fork, building it for netstandard2.0 and including into your project.
